container-search/src/main/java/com/yahoo/prelude/fastsearch/FS4InvokerFactory.java

```diff
@@ -43,11 +42,27 @@ public FS4InvokerFactory(FS4ResourcePool fs4ResourcePool, SearchCluster searchCl
     }
 
     public SearchInvoker getSearchInvoker(Query query, SearchCluster.Node node) {
-        return new FS4SearchInvoker(searcher, query, fs4ResourcePool, node);
+        Backend backend = fs4ResourcePool.getBackend(node.hostname(), node.fs4port());
+        return new FS4SearchInvoker(searcher, query, backend.openChannel(), node);
     }
 
     public Optional<SearchInvoker> getSearchInvoker(Query query, List<SearchCluster.Node> nodes) {
-        return getInvoker(nodes, node -> getSearchInvoker(query, node), InterleavedSearchInvoker::new);
+        Map<Integer, SearchInvoker> invokers = new HashMap<>();
+        for (SearchCluster.Node node : nodes) {
+            if (node.isWorking()) {
+                Backend backend = fs4ResourcePool.getBackend(node.hostname(), node.fs4port());
+                if (backend.probeConnection()) {
+                    invokers.put(node.key(), new FS4SearchInvoker(searcher, query, backend.openChannel(), node));
+                } else {
+                    return Optional.empty();
```

Comment:
We will never choose a group with any connectivity issues? Consider the case of two large groups with some minor connectivity issues such that one node in each group is down. And the case of just one node down and as a consequence the other group receiving all the queries and thus becoming overloaded.

I think that to do this correctly you need two passes, but then you start duplicating the logic where you make holistic group decisions in ping complete. Why did you want to make decisions based on connectivity here? Can't we just stick to the decisions made after pinging?

Reply:
I added the probing because in the system test waiting for the pings just wasn't responsive enough to behave correctly. Bailing out here happens when the connectivity fails right now and the query would effectively already fail. The loop in Dispatcher that retries other groups is maybe more suspect -- it tries *number-of-groups* times to find a connected group and tells SearchCluster to fail a group until the next ping iteration (in <1s, though).

Perhaps a better version would be to try at most N times (N being 2 or 3) and get that list from the LB in one go. The last choice could be used forcibly without regard for connectivity issues (if *some* result is better than *no* result). SearchCluster could also be extended to take the connection failure with a grain of salt and only fail the group once a threshold is hit. Assuming the connection failure is intermittent the queries-in-flight counter should take care of balancing the load.

Reply:
> The last choice could be used forcibly without regard for connectivity issues (if some result is better than no result)

Yes, that sounds good. I think some result is better than none from the point of view of this code since the ping logic determined that the group should be serving taking the coverage config into account.